matching/README.md

```diff
@@ -0,0 +1,991 @@
+no row in an upper group applies. This allows us to improve pattern matching
+efficiency since K does not, for the most part, case what order rules are tried
+in.
+
+Matrices are first simplified to remove or-patterns by replacing each row
+containing an or-pattern with a set of rows corresponding to all possible
+permutations of the disjuncts of all or-patterns in the pattern.
+
+### Decision Trees
+
+A decision tree is a tree whose root represents the input of the matching
+process and whose leaves identify which rewrite rule, if any, applies to the
+term being matched. It has the following BNF-style syntax:
+
+```
+syntax DT ::= "Failure"
+            | Success(Int, [(Occurrence, Sort)])
+            | Switch(Occurrence, Sort, [(Constructor, DT)], DT?)
+            | CheckNull(Occurrence, Sort, DT, DT)
+            | Function(Occurrence, Function, [(Occurrence, Sort)], Sort, DT)
+            | Pattern(Occurrence, Sort, Pattern, DT)
+            | IterHasNext(Occurrence, Sort, DT, DT)
+```
+
+### Heuristics
+
+The only heuristics regularly used by the LLVM backend are `q`, `b`, `a`, and
+the `L` pseudo-heuristic. We will discuss in a later section how these
+heuristics are adapted to the new features of the algorithm.
+
+## Compilation
+
+Here we describe the compilation algorithm, the updated specialization and
+default decompositions, and the adapted heuristics.
+
+### Initial matrix
+
+Starting from a set of rewrite rules of size n and their priorities, where each
+rewrite rule has a left-hand side and a right-hand side of sort s, and
+optionally a boolean condition Cond, we compute the initial matrix as follows:
+
+P starts with one column and n rows. P_i1 = LHS[i]. o_1 = []. s_1 = s.
+C_i = (a_i = i, S_i = empty, Cond_i = Cond[i], R_i = []).
+
+The rows of P are sorted in increasing priority order, where each set of rows
+with the same priority forms a group.
+
+### Compile function
+
+`Compile` is a function that takes a Matrix and returns a Decision Tree.
+
+Described below is pseudocode for the function:
+
+```
+def compile(Matrix m): DT
+  if m has no rows:
+    return Failure
+  else:
+    case bestRow(m) of
+      None -> decomposeBestCol(m)
+      BestRow(i) -> getLeaf(i, compile(m with row i removed))
+
+def bestRow(Matrix m): Int
+  if m has no columns:
+    return the first row of m
+  if there is a row in the topmost group of m where all columns are wildcard patterns:
+    return that row
+  else:
+    return None
+
+def getLeaf(Row i, DT dt): DT
+  val newVars = empty
+  for j = 1 to m:
+    for each variable X in P_ij:
+      newVars[X] = o_j
+  val allVars = newVars U S_i
+  val vars = sort the values of allVars by their keys alphabetically
+  val atomicLeaf = Success(a_i, vars)
+  if Cond_i is not None:
+    val f = a function that evaluates Cond_i
+    val sc = Function(Fresh(a_i), f, vars, Bool, dt)
+  else:
+    val sc = atomicLeaf
+  return foldRanges(R_i, dt)
```

Comment:
I don't understand the intended semantics of the side condition parts of the decision tree structure. The variable `sc` isn't used at all here, and the `atomicLeaf` isn't used in the part where we construct the `Function` term.

Reply:
whoops, you are correct. It's a typo.